session.go

```diff
@@ -419,14 +434,35 @@ func (s *Session) send() {
 				if err != nil {
 					s.logger.Printf("[ERR] yamux: Failed to write body: %v", err)
 					asyncSendErr(ready.Err, err)
+					s.sendDeadCh <- true
 					s.exitErr(err)
 					return
 				}
 			}
 
 			// No error, successful send
 			asyncSendErr(ready.Err, nil)
+		})
+
+		select {
+		case ready := <-s.sendCh:
+			handle(ready)
+
 		case <-s.shutdownCh:
+			done := false
+
+			for !done {
+				select {
+				case ready := <-s.sendCh:
+					handle(ready)
+
+				default:
+					done = true
+				}
+			}
+
+			s.sendDeadCh <- true
```

Comment:
`handle` will have already sent a value over sendDeadCh. Is it intentional to send another value here? I realize that it is probably "fine" because the channel is buffered and this probably won't cause the goroutine to block forever. But if that's not intentional I would rather not need to consider that possibility at all.

Reply:
I guess technically it's unknown if `handle` sent a value over sendDeadCh. But I still dislike the apparent ambiguity of sendDeadCh being sent between 0 and 2 values

Reply:
I think I see what you mean, and in fact sendDeadCh could be sent an arbitrary number of values as handle can be called multiple times. I'll look into whether it's possible to make handle return a `bool` indicating that the stream has been closed and it is not necessary to send to sendDeadCh or process further `sendCh` events. In practice gorouting leakage was not seen during testing but maybe there is a use pattern where it would show up.

Reply:
I also see that if channel close is used as the signal, it doesn't matter if it's sent multiple times.

Reply:
Ah. Since you have found that channels can only be closed once perhaps sync.Once would be of help?